README.md
[![Build Status](https://travis-ci.com/chandojo/ExerciseCoachTools.svg?branch=master)](https://travis-ci.com/chandojo/ExerciseCoachTools)

# Exercise Coach Tools

Exercise Coach Tools is an open source library of tools for fitness/exercise coaches, fitness specialists, exercise physiologists, and personal trainers.  The objective of Exercise Coach Tools is to help these people be more efficient and spend more time focusing on their clients instead of the equations/formulas needed to make analyses.

## Table of Contents
1. [Getting Started](#gettingstarted)
2. [ACSM Risk Stratification](#acsmriskstratification)
3. [Percentage Calculators](#percentagecalculators)
4. [Metabolic Rate Calculators](#metabolicratecalculators)
5. [Workout Programs](#workoutprograms)
6. [Testing](#testing)

## Getting Started <a name="gettingstarted"></a>
### Prerequisites

This package requires **>= Python 3.7.4**

### Installing

`pip install exercisecoachtools`

## ACSM Risk Stratification <a name="acsmriskstratification"></a>
This is a tool for determining risk of cardiovascular disease based on *ACSM’s Guidelines for Exercise Testing and Prescription-8th ed. Philadelphia: Lippincott Williams & Wilkins*

This tool is not a substitute for advice from a doctor and is used to help fitness specialists refer potential clients to physicians due to possible cardiovascular health risk.

### Using the Stratification
A `Patient` object must be created to use methods.

`Patient` takes the following parameters in the order listed:
- sex (str *male or female*)
- age (int *positive only*)
- smoker (str *yes or no*)
- sedentary (str *yes or no*)
- bmi (str *yes or no* ; input *0* if no available value but waist_girth value is available )
- waist_girth (int *positive only* ; input *0* if no available value but bmi value is available)
- male_family_death_before_55 (str *yes or no*)
- female_family_death_before_65 (str *yes or no*)
- systolic (int *positive only*)
- diastolic (int *positive only*)
- hypertensive (str *yes or no*)
- ldl (int *positive only*)
- hdl (int *positive only*)
- using_lipid_lowering_medication (str *yes or no*)
- cholesterol (int *positive only*)
- fasting_glucose (int *positive only* ; input *0* if no available value but oral_glucose_tolerance value is available)
- oral_glucose_tolerance (int *positive only* ; input *0* if no available value but fasting_glucose value is available)

**Useful functions:**
- `risk_classification(Patient)`
  - Returns patient's risk classification based off of the ACSM Guidelines. Classifications range from low, moderate, and high.
- `total_risk_factors(Patient)`
  - Returns patient's net total risk factors. This total is based off of ACSM's algorithm.


#### Examples
```
from risk.risk_stratification import Patient, risk_classification, total_risk_factors

steve_buscemi = Patient('male', 61, 'no', 'yes', 25, 0, 'yes', 'no', 119, 78, 'no', 100, 70, 'no', 100, 60, 0)

risk_classification(steve_buscemi)
# 'Your risk total is 2. You are at a moderate risk for cardiovascular disease. Medical check-up recommended for participation in vigorous physical activity.'

total_risk_factors(steve_buscemi)
# returns 2

```

## Percentage Calculators <a name="percentagecalculators"></a>

**Useful functions:**
- `percentage_value(desired_percentage, 1_rep_max_weight)`
  - Returns weight value associated with desired percentage. Can use kg or lbs
- `one_rep_max_estimation(reps completed, weight)`
  - Returns dictionary of estimated 1-rep max using Brzycki, Epley, Mayhew, Lander, and Lombardi formulas
  - Note: Reps completed **must be** > 1

#### Examples

```
from percentage_calculators.percentage_calculators import percentage_value

percentage_value(80, 100)
# returns 80.0
```

```
from percentage_calculators.one_rep_max_estimation import one_rep_max_estimation

one_rep_max_estimation(2, 100)
# returns {'brzycki_formula': 102.85949393128985, 'epley_formula': 106.66666666666667, 'mayhew_formula': 106.41304152560984, 'lander_formula': 104.21275910157765, 'lombardi_formula': 107.17734625362931}
```

## Metabolic Rate Calculators <a name="metabolicratecalculators"></a>

A `MetabolicClient` object must be created to use methods.

`MetabolicClient` takes the following parameters in the order listed:
- sex (str *male or female*)
- age (int *positive only*)
- weight (pounds - float *positive only*)
- height (inches - float *positive only*)

**Useful functions:**
- `bmr_results(MetabolicClient)`
  - Returns client's estimated BMR using Harris-Benedict and Mifflin formulas
- `harris_benedict_bmr(MetabolicClient)`
  - Returns client's estimated BMR using Harris-Benedict formula
- `mifflin_bmr(MetabolicClient)`
  - Returns client's estimated BMR using Mifflin St Jeor formula

- `get_calories_by_activity(activity_level, bmr)`
  - Returns client's daily metabolic rate based on activity level
  - Activity level options:
      - Sedentary — desk job and little to no exercise
      - Lightly Active — light exercise/sports 1–3 days/week
      - Moderately Active — moderate exercise/sports 3–5 days/week
      - Very Active — hard exercise/sports 6–7 days/week
      - Extremely Active — hard daily exercise/sports and physical job or training


#### Examples
```
from kcal.metabolic_rate_prediction import MetabolicClient, bmr_results, harris_benedict_bmr, mifflin_bmr
from kcal.daily_kcal_by_lifestyle import get_calories_by_activity

sue_bird = MetabolicClient('female', 39, 150, 69)

bmr_results(sue_bird)
# returns {'harris_benedict_bmr': 974.5799999999999, 'mifflin_st_jeor_bmr': 826.6300000000001}

sue_bird_bmr = harris_benedict_bmr(sue_bird)
# returns 974.5799999999999

mifflin_bmr(sue_bird)
# returns 826.6300000000001

get_calories_by_activity('very active', sue_bird_bmr)
# returns 1681.1505 (kcal/day)
```

## Workout Programs <a name="workoutprograms"></a>

### Current workout programs:
- `fivethreeone_program` : Returns [weight, reps] to use each week per set. Based off of Jim Wendler's 5/3/1 program.  For a description of the program read [here](https://www.t-nation.com/workouts/531-how-to-build-pure-strength). 


#### Examples
```
from workout_programs import Workouts

my_workout= Workouts(100)

my_workout.fivethreeone_program()
# returns {'Week 1': [[58.5, 5], [67.5, 5], [76.5, 5]], 'Week 2': [[63.0, 3] [72.0, 3], [81.0, 3]], 'Week 3': [[67.5, 5], [76.5, 3], [85.5, 1]], 'Week 4': [[36.0, 5], [45.0, 5], [54.0, 5]]}

```

## Running the tests <a name="testing"></a>
`python setup.py test`

## Authors
chandojo


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]
### Added
- Workout programs module. Includes workouts:
  - 5/3/1

### Changed
- Added `.pytest_cache` to `.gitignore`

### Fixed
- CI/CD deployment only on `master` branch  

## [0.1.2] - 2020-09-07
### Changed
- Added continuous delivery to PYPI in TravisCI configuration

## [0.1.1] - 2019-12-16
- Initial release













tests/test_max_percentage.py
import unittest
from percentage_calculators.percentage_calculators import *


class MaxPercentageTestMethods(unittest.TestCase):
    def test_max_percentage(self):
        answer = 80
        return_value = percentage_value(80, 100)
        self.assertEqual(answer, return_value)


tests/test_workoutprograms.py
import unittest
from workout_programs import Workouts


class WorkoutProgramsTestMethods(unittest.TestCase):
    def test_percentages(self):
        expected_value = 65
        return_value = Workouts(100)._get_percentages()
        self.assertEqual(expected_value, return_value[65])

    def test_fivethreeone_program(self):
        expected_value = 63
        return_value = Workouts(100).fivethreeone_program()
        self.assertEqual(expected_value, return_value["Week 2"][0][0])

tests/test_value_validation.py
import unittest
from validation.value_validation import *

class ValidateValuesSuccessTestMethods(unittest.TestCase):
    def test_valueIsMaleFemale_success(self):
        person = "male"
        valueMaleFemale = ValidateValue(person).valueIsMaleFemale()
        self.assertEqual(person, valueMaleFemale)

    def test_valueIsPositiveInteger_success(self):
        number_value = 100
        valuePositiveInteger = ValidateValue(number_value).valueIsPositiveInteger()
        self.assertEqual(number_value, valuePositiveInteger)

    def test_valueIsPositiveFloat_success(self):
        number_value = 88.88
        valuePositiveFloat = ValidateValue(number_value).valueIsPositiveFloat()
        self.assertEqual(number_value, valuePositiveFloat)

    def test_valueYesNo_success(self):
        valueYesNo = ValidateValue('yes').valueIsYesNo()
        self.assertEqual(True, valueYesNo)


class ValidateValuesFailsTestMethods(unittest.TestCase):
    def test_valueIsMaleFemale_fails(self):
        kitten = ValidateValue('kitten')
        self.assertRaises(ValueError, kitten.valueIsMaleFemale)

    def test_valueIsPositiveInteger_fails(self):
        negative = ValidateValue(-100)
        word = ValidateValue('hello')
        self.assertRaises(ValueError, negative.valueIsPositiveInteger)
        self.assertRaises(ValueError, word.valueIsPositiveInteger)

    def test_valueIsPositiveFloat_success(self):
        negative = ValidateValue(-99.143)
        word = ValidateValue('boop')
        self.assertRaises(ValueError, negative.valueIsPositiveFloat)
        self.assertRaises(ValueError, word.valueIsPositiveFloat)

    def test_valueIsYesNo_fails(self):
        bloop = ValidateValue('bloop')
        self.assertRaises(ValueError, bloop.valueIsYesNo)


risk/risk_stratification.py
from validation.value_validation import ValidateValue

class Patient(object):
    def __init__(self, sex, age, smoker, sedentary, bmi, waist_girth, male_family_death_before_55, female_family_death_before_65, systolic, diastolic, hypertensive, ldl, hdl, using_lipid_lowering_medication, cholesterol, fasting_glucose, oral_glucose_tolerance):
        self._sex = ValidateValue(sex).valueIsMaleFemale()
        self._age = ValidateValue(age).valueIsPositiveInteger()
        self._smoker = ValidateValue(smoker).valueIsYesNo()
        self._sedentary = ValidateValue(sedentary).valueIsYesNo()
        self._bmi = ValidateValue(bmi).valueIsPositiveInteger()
        self._waist_girth = ValidateValue(waist_girth).valueIsPositiveInteger()
        self._male_family_death_before_55 = ValidateValue(male_family_death_before_55).valueIsYesNo()
        self._female_family_death_before_65 = ValidateValue(female_family_death_before_65).valueIsYesNo()
        self._systolic = ValidateValue(systolic).valueIsPositiveInteger()
        self._diastolic = ValidateValue(diastolic).valueIsPositiveInteger()
        self._hypertensive = ValidateValue(hypertensive).valueIsYesNo()
        self._ldl = ValidateValue(ldl).valueIsPositiveInteger()
        self._hdl = ValidateValue(hdl).valueIsPositiveInteger()
        self._using_lipid_lowering_medication = ValidateValue(using_lipid_lowering_medication).valueIsYesNo()
        self._cholesterol = ValidateValue(cholesterol).valueIsPositiveInteger()
        self._fasting_glucose = ValidateValue(fasting_glucose).valueIsPositiveInteger()
        self.oral_glucose_tolerance = ValidateValue(oral_glucose_tolerance).valueIsPositiveInteger()


class RiskFactorAssessment(object):
    def __init__(self, patient):
        self._patient = patient

    def result_risk_classification(self):
        classification = RiskFactorAssessmentClassification().patient_risk_classification(self.net_risk_factors())
        return classification

    def net_risk_factors(self):
        total = self.get_risk_factor_count() - self.get_negative_risk_factor_count()
        return total

    def get_risk_factor_count(self):
        _count_risk_factors = [
            self._is_age_risk(),
            self._is_obesity_risk(),
            self._patient._smoker,
            self._patient._sedentary,
            self._is_familial_risk(),
            self._is_systolic_risk(),
            self._is_diastolic_risk(),
            self._patient._hypertensive,
            self._is_dyslipidemia_risk(),
            self._is_pre_diabetes_risk()
            ]
        return _count_risk_factors.count(True)

    def get_negative_risk_factor_count(self):
        _count_negative_risk_factor = [ self._is_hdl_negative_risk() ]
        return _count_negative_risk_factor.count(True)

    def _is_obesity_risk(self):
        if self._patient._bmi and self._patient._waist_girth == 0:
            return False
        elif self._patient._bmi > 30:
            return True
        elif self._patient._waist_girth > 40 and self._patient._sex == 'male' or self._patient._waist_girth > 35 and self._patient._sex == 'female':
            return True
        else:
            return False

    def _is_age_risk(self):
        if (self._patient._sex == "male" and self._patient._age >= 45) or (self._patient._sex == "female" and self._patient._age >=55):
            return True
        else:
            return False

    def _is_familial_risk(self):
        if self._patient._male_family_death_before_55 == True or self._patient._female_family_death_before_65 == True:
            return True
        else:
            return False

    def _is_systolic_risk(self):
        if self._patient._systolic >= 120:
            return True
        else:
            return False

    def _is_diastolic_risk(self):
        if self._patient._diastolic >= 80:
            return True
        else:
            return False

    def _is_dyslipidemia_risk(self):
        if self._patient._ldl > 130 or self._patient._hdl < 40 or self._patient._cholesterol > 200:
            return True
        else:
            return False

    def _is_pre_diabetes_risk(self):
        if self._patient._fasting_glucose >= 100 and self._patient._fasting_glucose <= 126 or self._patient._oral_glucose_tolerance >= 140 and self._patient._oral_glucose_tolerance < 200:
            return True
        else:
            return False

    def _is_hdl_negative_risk(self):
        if self._patient._hdl > 60:
            return True
        else:
            return False

class RiskFactorAssessmentClassification(object):
    def __init__(self):
        pass

    def patient_risk_classification(self, value):
        if value <= 1:
            return self._low_risk_category(value)

        if value == 2:
            return self._moderate_risk_category(value)

        if value >  2:
            return self._high_risk_category(value)

    def _low_risk_category(self, value):
        return("Your risk total is %s. You are at a low risk for cardiovascular disease. Medical check-up no necessary for participation in physical activity." % value)

    def _moderate_risk_category(self, value):
        return("Your risk total is %s. You are at a moderate risk for cardiovascular disease. Medical check-up recommended for participation in vigorous physical activity." % value)

    def _high_risk_category(self, value):
        return("Your risk total is %s. You are at a high risk for cardiovascular disease. Medical check-up highly recommended before any physical activity." % value)


validation/value_validation.py
class ValidateValue():
    _female_male = ['female', 'male']
    _yes_no = {
                'yes': True,
                'no': False
                }

    def __init__(self, value):
        self._value = value

    def valueIsMaleFemale(self):
        if self._value in ValidateValue._female_male:
            return self._value
        else:
            raise ValueError('%s is an incorrect value. It must be male or female.' % self._value)

    def valueIsPositiveInteger(self):
        integer_value = int(self._value)
        if integer_value > 0:
            return integer_value
        else:
            raise ValueError('%s is an incorrect value. It must be a number and must higher than zero.' % self._value)

    def valueIsPositiveFloat(self):
        float_value = float(self._value)
        if float_value > 0:
            return float_value
        else:
            raise ValueError('%s is an incorrect value. It must be a number and must higher than zero.' % self._value)

    def valueIsYesNo(self):
        if self._value in ValidateValue._yes_no.keys():
            return ValidateValue._yes_no[self._value]
        else:
            raise ValueError('%s is an incorrect value. Must be yes or no.' % self._value)


workout_programs.py
class Workouts():
    def __init__(self, max):
        self._max = max

    def _get_percentages(self):
        percentages = {}

        i = .65
        while i < 1:
            value = round(i * self._max, 2)
            key = int(i*100)
            percentages[key] = value
            i = i + .05

        d = .40
        while d < .70:
            value = round(d * self._max, 2)
            key = int(d*100)
            percentages[key] = value
            d = d + .10

        return percentages
    
    
    def fivethreeone_program(self):
        percentages = self._get_percentages()
        
        for key, val in percentages.items():
            percentages[key] = val * .9

        workout = {
            "Week 1": [
                [percentages[65], 5],
                [percentages[75], 5],
                [percentages[85], 5]
            ],
            "Week 2": [
                [percentages[70], 3],
                [percentages[80], 3],
                [percentages[90], 3]
            ],
            "Week 3": [
                [percentages[75], 5],
                [percentages[85], 3],
                [percentages[95], 1]
            ],
            "Week 4": [
                [percentages[40], 5],
                [percentages[50], 5],
                [percentages[60], 5]
            ]
        }

        return workout